File: python/graphs/word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        adjacency_list = {}
        words = wordList + [beginWord]
        for word in words:
            for j in range(len(word)):
                adjacency_list[word[:j] + "*" + word[j + 1:]] = adjacency_list.get(word[:j] + "*" + word[j + 1:], [])
                adjacency_list[word[:j] + "*" + word[j + 1:]].append(word)

        q = deque()
        q.append((beginWord, 1))
        visited = set()
        while q:
            current = q.popleft()

            if current[0] in visited:
                continue

            if current[0] == endWord:
                return current[1]

            visited.add(current[0])

            for i in range(len(current[0])):
                for neighbor in adjacency_list[current[0][:i] + "*" + current[0][i + 1:]]:
                    q.append((neighbor, current[1] + 1))

        return 0

File: c++/graphs/word-ladder.cc
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        std::unordered_map<std::string, std::vector<std::string>> ladder;

        wordList.push_back(beginWord);
        for (std::string word : wordList) {
            for (int i = 0; i < word.size(); ++i) {
                std::string wildcard = word;
                wildcard[i] = '.';
                ladder[wildcard].push_back(word);
            }
        }
        
        std::unordered_set<std::string> visited;
        std::queue<std::pair<std::string, int>> q;
        q.push(std::make_pair(beginWord, 1));
        while (q.size() > 0) {
            std::pair<std::string, int> current = q.front();
            q.pop();

            if (visited.contains(current.first)) {
                continue;
            }

            visited.insert(current.first);

            if (current.first == endWord) {
                return current.second;
            }

            for (int i = 0; i < current.first.size(); ++i) {
                std::string wildcard = current.first;
                wildcard[i] = '.';
                for (std::string neighbor : ladder[wildcard]) {
                    q.push(std::make_pair(neighbor, current.second + 1));
                }
            }
        }

        return 0;
    }
};
